packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,156 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+(String? resolutionPluginName, String? defaultImplementation) _getPluginImplementationCandidate(Plugin plugin, String platformKey) {
+  String? defaultImplementation = plugin.defaultPackagePlatforms[platformKey];
+  if (plugin.platforms[platformKey] == null && defaultImplementation == null) {
+    // The plugin doesn't implement this platform.
+    return (null, null);
+  }
+  String? implementsPackage = plugin.implementsPackage;
+  final bool hasInlineDartImplementation =
+      plugin.pluginDartClassPlatforms[platformKey] != null &&
+          plugin.pluginDartClassPlatforms[platformKey] != 'none';
+  if (implementsPackage == null || implementsPackage.isEmpty) {
+    if (defaultImplementation == null && !hasInlineDartImplementation) {
+      // Skip native inline PluginPlatform implementation
+      return (null, null);
+    }
+    if (defaultImplementation != null) {
+      return (null, defaultImplementation);
+    } else {
+      // An app-facing package (i.e., one with no 'implements') with an
+      // inline implementation should be its own default implementation.
+      // Desktop platforms originally did not work that way, and enabling
+      // it unconditionally would break existing published plugins, so
+      // only treat it as such if either:
+      // - the platform is not desktop, or
+      // - the plugin requires at least Flutter 2.11 (when this opt-in logic
+      //   was added), so that existing plugins continue to work.
+      // See https://github.com/flutter/flutter/issues/87862 for details.
+      final bool isDesktop = platformKey == 'linux' || platformKey == 'macos' || platformKey == 'windows';
+      final semver.VersionConstraint? flutterConstraint = plugin.flutterConstraint;
+      final semver.Version? minFlutterVersion = flutterConstraint != null &&
+          flutterConstraint is semver.VersionRange ? flutterConstraint.min : null;
+      final bool hasMinVersionForImplementsRequirement = minFlutterVersion != null &&
+          minFlutterVersion.compareTo(semver.Version(2, 11, 0)) >= 0;
+      if (!isDesktop || hasMinVersionForImplementsRequirement) {
+        // Given: implementsPackage == null, defaultImplementation == null, hasInlineDartImplementation == true
+        implementsPackage = plugin.name;
+        defaultImplementation = plugin.name;
+        return (implementsPackage, defaultImplementation);
+      } else {
+        // If it doesn't meet any of the conditions, it isn't eligible for
+        // auto-registration.
+        return (null, null);
```

Comment:
For the other return statements, there's no case, where `defaultImplementations` is assigned and not returned, so can return always null for that.